attribute-service-impl/src/main/java/org/hypertrace/core/attribute/service/delegate/AttributeUpdaterImpl.java

```diff
@@ -0,0 +1,93 @@
+package org.hypertrace.core.attribute.service.delegate;
+
+import static java.util.stream.Collectors.toUnmodifiableList;
+import static org.hypertrace.core.documentstore.expression.impl.LogicalExpression.and;
+import static org.hypertrace.core.documentstore.model.options.UpdateOptions.DEFAULT_UPDATE_OPTIONS;
+
+import io.grpc.Status;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+import org.hypertrace.core.attribute.service.AttributeMetadataValidator;
+import org.hypertrace.core.attribute.service.builder.AttributeFilterBuilder;
+import org.hypertrace.core.attribute.service.builder.AttributeFilterBuilderImpl;
+import org.hypertrace.core.attribute.service.builder.AttributeUpdateBuilder;
+import org.hypertrace.core.attribute.service.builder.AttributeUpdateBuilderImpl;
+import org.hypertrace.core.attribute.service.converter.AttributeMetadataConverter;
+import org.hypertrace.core.attribute.service.converter.AttributeMetadataConverterImpl;
+import org.hypertrace.core.attribute.service.v1.UpdateMetadataRequest;
+import org.hypertrace.core.attribute.service.v1.UpdateMetadataResponse;
+import org.hypertrace.core.documentstore.Collection;
+import org.hypertrace.core.documentstore.Document;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.model.subdoc.SubDocumentUpdate;
+import org.hypertrace.core.documentstore.query.Query;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+public class AttributeUpdaterImpl implements AttributeUpdater {
+  private final Collection collection;
+  private final AttributeMetadataConverter converter;
+  private final AttributeFilterBuilder filterBuilder;
+  private final AttributeUpdateBuilder updateBuilder;
+
+  public AttributeUpdaterImpl(final Collection collection) {
+    this.collection = collection;
+    this.converter = new AttributeMetadataConverterImpl();
+    this.filterBuilder = new AttributeFilterBuilderImpl();
+    this.updateBuilder = new AttributeUpdateBuilderImpl();
+  }
+
+  @Override
+  public UpdateMetadataResponse update(
+      final UpdateMetadataRequest request, final RequestContext context) throws IOException {
+    final String tenantId = AttributeMetadataValidator.validateContextAndGetTenantId(context);
+    validate(request);
+
+    final FilterTypeExpression filter = buildFilter(request, tenantId);
+    final Query query = Query.builder().setFilter(filter).build();
+    final List<SubDocumentUpdate> updates = buildUpdates(request);
+    final Optional<Document> docOptional =
+        collection.update(query, updates, DEFAULT_UPDATE_OPTIONS);
```

Comment:
I might be missing it, but the tricky bit about updating here is the hierarchy which looks like it's missing. If I try to update a root-owned attribute, we should allow it, but we can't update the actual root attribute, otherwise we'd leak to other tenants. We need to create a clone that applies the update for the requesting tenant.

If I'm reading the code in this PR right, we're only supporting updates on attributes that are already defined in the requesting tenant's context. That may be sufficient for first cut, but at a minimum we should call it out.

Reply:
Good point, Aaron. Will update it.

Reply:
Updated the message `No attribute found with id: %s.` to `No custom attribute found with id: %s.`
I don't think we need to make an additional DB call to differentiate the actual non-existence of an attribute vs the existence of an attribute at the root level and supply the message accordingly. Do you agree?

Reply:
Yes. If we're going to put in the effort to check existence, might as well add the support for updating attributes that aren't custom.